### DISCUSSION TITLE
Welcome to Tonuino-ESP32-I2S Discussions!

### DISCUSSION THREAD

**Discussion by biologist79** [biologist79/ESPuino#24] [General]
Starting today, discussions are available on my project. Feel free to contribute/ask in German or English.
Heute geht's los mit einem Forum hier. Schreibt mir gerne und diskutiert. Egal ob in Englisch oder in Deutsch.

**Comment by Soehm** [comment]
Hey Biologist,

ich verfolge dein Projekt hier schon etwas länger und bereite mich gerade auf eine ESP32 NodeMCU / DevKitC Variante vor.
Ich habe gesehen, das es bereits ein "ESP32 Breakout Board" gibt - nicht von dir, ist mir klar :-)

Zu den Unterschieden habe ich trotzdem 2 Fragen:

- Deine Variante mit Lolin32 benötigt am RFC522 keinen RTS Pin, der DevKitC schon. Gibt es da eine einfache Erklärung zu?
- Benötigt die "Headphone-Detection" (GPIO22) den I2C oder reicht ein Eingangssignal und ich könnte es auf GPIO35 umlegen?

Danke für deine Zeit und alles Gute!
Stefan

**Comment by biologist79** [comment]
Moin,
ist korrekt: Das Board stammt nicht von mir und ich habe es auch nicht getestet. Ich statte meine Tonuinos inzwischen hat auch mit Akkus aus und da eignen sich die Lolins einfach besser für.

Zu deinen Fragen:

zu 1) Das kann ich dir ehrlich gesagt nicht beantworten. Ich hatte das immer im Setup SPI + RST laufen. Bis ich dann gemerkt habe, dass man RST gar nicht benötigt. Da habe ich es einfach weg gelassen. Läuft dennoch. Und zwar an einem NodeMCU, Lolin32, LolinD32, Lolin D32 pro und Audiokit. Ich bin, das mal als Preview, aktuell noch dabei, für den Lolin32 ein Board zu machen, welches als RFID-Reader den PN5180 (bessere Reichweite) und statt SPI für die SD-Anbindung MMC verwendet (ca. 310 kB/s bei FTP). Aber hab erst den Schaltplan gezeichnet und muss es noch erst routen.
zu 2) Ich hab's jetzt nicht getestet, aber das müsstest du problemlos umbelegen können (i2c wird nicht benötigt, das ist nur Zufall, dass es auf i2c liegt). Man muss halt bedenken, dass die "höheren" GPIOs eigentlich nur GPIs sind und auch keinen internen Pullup haben. Aber das macht ja hier nix.

VG
Torsten.

**Comment by Soehm** [comment]
Hey Torsten,

danke für deinen Input. Die Antwort zu 1 erübrigt eigentlich die Frage 2. Dann kann ich die Headphone Detection ja vermutlich auf PIN22 lassen.

Nutzt du nur KiCad oder hast du die (Preview)Zeichnung auch bei Easyeda? Ich habe dein Layout und die Kopfhörerplatine von compactflash zu easyeda portiert und wollte beides auf einer Platine "voll" bestückt bestellen. Die Kopfhörerplatine sprengt aber den Rahmen, da die Chips (und auch vergleichbare) nicht zu bekommen oder aber äußerst teuer sind/werden.

Stand jetzt habe ich mich 

1. zu Testzwecken für dein Layout + Erweiterung um einen Trägerteil für den PCM5102a + TDA1308 (beide als fertige Platinen zum auflöten) entschieden 
2. parallel dazu eine AiO Platine mit "Custom designed" ESP32 entschieden.

1. ist nahezu fertig - nach deinem Input
2. ist bei 75% Zeichnung fertig - ich traue mich aber erst, wenn 1. funktioniert :-)

Mal schauen, wo es sich hin "entwickelt"...

Beste Grüße
Stefan

**Reply by biologist79** [reply]
Ich arbeite mit KiCad, aber auch noch nicht lange. Ich habe ja hier gut 3/4 Jahr erstmal keine PCBs gehabt, da es für mich eine Streifenrasterplatine auch getan hat. Aber ich wurde immer öfter danach gefragt und dann habe ich mich halt mal hingesetzt und KiCad "gelernt" :-) Ist halt auch schon echt grandios, für wie günstig man den Kram fertigen lassen kann. Da kann ich ja nicht mal Streifenrasterplatinen für kaufen.

Das Schöne an der Kopfhörerplatine ist halt, dass sie eine Buchse besitzt, die erkennt, ob was eingesteckt ist. Das ist z.B. beim UDA1334, welches ich auch hier habe, anders: Die funktioniert gut (ok, ist eigentlich auch eher für Line als für Kopfhörer gedacht), ist in China billig und auch kompakt. Hat aber halt keine schaltbare Buchse. Was sie aber immerhin besitzt, sind rausgeführte analoge Ausgänge. Da könnte man sich natürlich auch einen schaltbaren Kopfhörerausgang dranmachen. Du beachten ist allerdings: Den MAX98357a schaltet man mit GND am SD-Button auf mute, während die Kopfhörerbuchse jedoch GND liefert, wenn nix eingesteckt ist. Also genau andersrum, als man es braucht. Daher hat's compactflash mit einem P-Channel-Mosfet invertiert.

Klingt auf jeden Fall gut mit der AiO. Compactflash hat schon verlauten lassen, dass er so richtig seinen Seelenfrieden nach wie vor nicht hat: Er würde gerne nicht mal mehr ein ESP32-Develboard verwenden und stattdessen einen ESP32-WROVER direkt auf eine AiO verbauen. Gut, Referenz-Designs kriegt man ja von Espressif, kann man schon machen. Die meisten Develboards setzen halt den WROOM ein. Beim WROVER hat man den Vorteil, dass er PSRAM besitzt, welcher z.B. beim Hören von Webradio als Cache verwendet wird (bei SD ist's egal). Vielleicht wird künftig auch PSRAM von mir noch mehr verwendet. Beispielsweise für Bluetooth (a2dp-sink). Aber das weiß ich noch nicht genau. Primär werde ich versuchen, es möglichst kompatibel für alle ESP32s zu halten.

VG
Torsten.

**Comment by Brain-san** [comment]
Hallo Torsten,

vielen Dank für das tolle Projekt. Ich hatte schon einen Tonuino gebaut und bin dann auf den Projekt gestoßen.
Plans du in Zukunft auch das anschließen eines Display? Ein kleines 128x32 zum Beispiel? Der Tonuino mit dem Nano und dem DFRobot Mini ist ja dahingehend beschränkt, dass es nicht direkt auf die SD-Karte zugreifen kann. Bei deiner Umsetzung kann man ja sogar die MP3-ID auslesen. Habe beim Tonuino festgestellt, dass meine Kinder blind die Knöpfe drücken, bis sie den richtigen Titel (Aus der gleichen Hörbuchreihe) gefunden haben. 
Wenn du das selber in Zukunft nicht vorhast, könntest du mich in die richtige Richtung schicken, welche Pins man verwenden müsste, da ja schon viele vom RFID und SD-Karten belegt sind.
Mit freundlichem Gruß
Gabor

**Comment by biologist79** [comment]
Hi,

also ich persönlich benötige ein Display nicht, aber es gab hier schon Ansätze, dies zu implementieren. Ich sag mal im Groben ist es auch kein Problem: Die Displays werden üblicherweise mit i2c angeschlossen, d.h. du brauchst halt zwei GPIOs. Musst halt drauf achten, dass es GPIOs <= (ich glaube) 34 sind, weil die höheren GPIOs taugen nur für Input (nicht aber für Output).
ESP32 hat ich sage mal Default-GPIOs für i2c, die verwendet werden, wenn du es nicht anders definierst. Es hindert dich aber niemand dran, andere GPIOs zu verwenden, wenn du das möchtest. 

Man braucht für i2c die Wire.h und instanziert dann mit Wire.begin(). Diese Funktion ist überladen, so dass du auch bei Bedarf dort die beiden Custom-Pins für SDA/SCL definieren kannst. Zu wissen gibt es ansonsten halt, dass du nicht nach Belieben (zu jedem Zeitpunkt) an die i3d-Tags drankommst. Ganz unten, am Ende von main.cpp, gibt es eine Reihe von Methoden, die bei Bedarf von der Audiolib aufgerufen werden (auf den Zeitpunkt deren Aufrufe hast du keinen Einfluss). Denke mal `void audio_id3data(const char *info) {` ist die richtige, aber weiß ich gerade nicht auswendig. Hier musst du char* halt ggf. in ein eigenes char* kopieren, wenn du zu beliebigen Zeitpunkten auf die Infos zugreifen willst.

**Reply by Brain-san** [reply]
Alles klar, dann muss ich mich mal daran setzen und versuchen, die Pin-Belegung anders zu definieren. Vielen Dank.

**Comment by biologist79** [comment]
Achso, zu den freien Pins: Hängt halt im Endeffekt davon ab, welche der Features du brauchst. Du musst dich auch nicht an die von mir vorgeschlagenen GPIOs halten. Sind halt quasi nur Vorschläge von mir.
Kannst zB auch hingehen, und die Buttons (prev, next, pplay) auf die "hohen" GPIOs legen, dann hast du die unteren für i2c frei. Allerdings werden für die Buttons Pullup-Widerstände in Hardware gebraucht, weil der ESP keine internen Pullups auf diesen GPIOs besitzt. Ist aber kein Hexenwerk: 10k Pullup auf 3.3V für jeden Button, der an 34-39 hängt und schon läuft das.

**Reply by tyllmoritz** [reply]
Hallo biologist79, 

erstmal vielen Dank für das tolle Projekt.
Auch an den Ersteller der Kopfhörerplatine und an tueddy für die Arbeiten mit dem PN5180.

Ich habe angefangen zu planen, wie ich mir einen Tonuino mit dem "Lolin D32 Pro" und dem PN5180 bauen kann.
Auf den Neopixel und den Rotary-Encoder würde ich gerne verzichten. Damit habe ich keinen Ein-Ausschalter.
Wenn Deep-Sleep funktioniert will ich aber auch keinen haben.

Mögliche Probleme bezüglich des Boot-Modus, wenn ich auf GPIO0 und GPIO12 Buttons lege, untersuche ich morgen.

Wenn ich mich etwas eingearbeitet habe, schaffe ich es vielleicht noch eine Platine für den "Lolin D32 Pro" zu erstellen.
Zum testen werde ich das aber ohne Platine versuchen.

hier ist mein Repo:
https://github.com/tyllmoritz/Tonuino-ESP32-I2S/tree/Lolin_D32_Pro

**Reply by biologist79** [reply]
Hi,

dann viel Spaß mit deinem Projekt :-)
GPIO0 habe ich bisher noch nie benutzt. Laut https://randomnerdtutorials.com/esp32-pinout-reference-gpios/ liegt da beim Booten ein PWM-Signal drauf und ein PullUp auf dem Eingang. Letztgenanntes ist gut, weil das brauchst du eh :-) Ist jetzt die Frage, wie lange da ein PWM-Signal drauf ist. Aber klingt gut. Aber was du zur Not, wenn dir die GPIOs ausgehen sollten auch machen kannst, ist über unterschiedliche Spannungsteiler die Buttons analog einzulesen. D.h. du hast einen analogen Eingang und je nach Widerstandswert, fällt eine andere Spannung ab. Das musst im Code dann halt zuordnen und dann weißt du, welche Taste gedrückt wurde.

Habe mich demletzt schon gefragt, was der D32 pro eigentlich mit den GPIOs 16 und 17 macht.

**Reply by tyllmoritz** [reply]
Der D32 nutzt die Pins für den PSRAM:
"ESP32’s GPIO16 and GPIO17 are used as chip select and clock signals for PSRAM." [0]

Ich habe mich noch weiter eingelesen und habe noch ein paar Fragen.
Ich bin leider noch nicht sehr erfahren, was Arduino/ESP32 angeht:

1. Warum wird für den PN5180 der Pin 22 (RFID_RST) mit HEADPHONE_DETECT geteilt?
2. Kann ich mit einem externen SD-Leser und dem Lolin D32 Pro auch den MMC-Modus verwenden?
    Mit dem internen SD-Slot geht das nicht, wie ich diesem Commit entnehmen kann [1].
    Ist mit 3 Pins mehr auch der 4-bit-Modus möglich, um die Geschwindigkeit der SD-Karte zu erhöhen?
3. Wenn ich den PN5180 als RFID-Reader benutze, könnte ich doch auf den SPI-Pins SCK, MOSI und MISO andere SPI-Geräte parallel anschließen, wenn ich den Cable-Select-Pin richtig benutze, oder?
   Damit müsste es möglich sein mit dem MCP23S17-E/SP [3] zusätzliche IO-Ports zu erlangen. (Für Neopixel, Rotary Encoder, LEDs, etc.)
   Der ESP32 ist für SPI der Master, oder? Kann ich für MISO (Master In, Slave Out) auch einen Pin >=34 benutzen?
4. Der ESP32 hat ja 2 I2S controller. Den zweiten Controller könnte man für SPDIF/Toslink nutzen. Ich plane jetzt erst einmal einen Pin zusätzlich für die Headphone-Platine zu reservieren… eventuell kann man ja irgendwann diesen Photolink-Transmitter integrieren [4].

[0] https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-wrover-kit.html
[1] https://github.com/biologist79/Tonuino-ESP32-I2S/commit/d19a3eb4e3d74cfa738dfd0cf3e0ca2120b867de#diff-d4b7b309d4e0bf9a92bb96bfab92e6620c4d844dcada0748c74b083059e6ac1d
[3] https://www.digikey.de/en/products/detail/microchip-technology/MCP23S17-E-SP/894276
[4] https://www.digikey.de/product-detail/de/everlight-electronics-co-ltd/EAPLTAA4/1080-1524-ND/7926541

**Reply by biologist79** [reply]
PSRAM: Ist korrekt. Lokal habe ich dazu schon was geschrieben, aber noch keinen Commit gemacht. Muss gestern, dass ich das bis vor ein paar Tagen nicht wusste, dass die für PSRAM verwendet werden und hab's (ungetestet) einfach hingeschrieben.

1) Beim D32 pro kriegt man einfach nicht alles zusammen hin, weil die GPIOs irgendwann "aus" sind. Es hat auch nicht jeder einen Kopfhöreranschluss und wenn, dann braucht vielleicht auch nicht jeder eine Lautstärke-Adaption, wenn ein Kopfhörer eingesteckt ist. Und ich habe oben ja geschrieben im File, dass ich es nur mit RC522 getestet habe. Also kurzum: Ich kann das nicht in allen Kombinationsmöglichkeiten ständig testen. Wenn ich Zeit habe, probiere ich mal wieder was aus, aber das muss nicht zwangsweise in allen Kombinationsmöglichkeiten zusammen laufen. So lange man was nicht mit einkompiliert ist es ja auch kein Problem, wenn was doppelt belegt ist.
2) Ja, das müsste eigentlich gehen. Im 4-bit-Modus habe ich das noch nie getestet. Wie ich gelesen habe, ist der Geschwindigkeitzuwachs, gemessen an dem, wieviel GPIOs man extra braucht, eher überschaubar. Da die Pins eh schon mehr oder weniger knapp sind, habe ich es gelassen. Wenn du MMC nutzt, dann denke dran, den PullUp-Widerstand an GPIO2 am SD-Reader zu entfernen. Weil sonst geht der ESP nicht in den Flashmodus, wenn man flashen will.
3) Theoretisch ja. Praktisch wirst du es testen und ggf. auch den Code anpassen müssen. Mit SINGLE_SPI hatte ich immer wieder Probleme (kommt keine Fehlermeldung aber beim RC522 passiert nix). Daher nutze ich das nicht mehr. Eine weitere Möglichkeit GPIOs zu sparen ist, die Buttons analog einzulesen über unterschiedliche Spannungsteiler. Damit könntest du zB alle drei Buttons und auch den Button vom Drehencoder über einen GPIO einlesen. Wenn jeder Button einen anderen Widerstand hat, kann man sie am Ende durch unterschiedliche Spannungswerte auseinander halten. Habe ich aber selbst noch nicht gemacht; muss man den Code anpassen. Vermutlich gibt's da auch "best practices", wie man die Widerstandswerte am besten wählt, damit man sie sicher diskriminieren kann. Hinweis: Man kann, wenn WiFi aktiv ist, auch nur manche GPIOs für analogRead() verwenden.
4) Ich hätte jetzt keine Verwendung für, aber wenn du das gebrauchen kannst - nurzu 👍. Wobei ich noch eine Anmerkung dazu habe: Wegen Fastled gab es zuweilen Absturzprobleme, weswegen ich auch schon am überlegen war, auf eine andere Lib zu wechseln. Diese würde i2s verwenden, um den Neopixel anzusteuern. Aber keine Ahnung, ob ich das wirklich machen werde. Das ganze Projekt kostet mich eh schon viel mehr Zeit, als ich es jemals gedacht hatte :-)

**Comment by tyllmoritz** [comment]
Hallo zusammen,


die Hardware von diesen Projekten sieht auch sehr vielversprechend aus:
Vermutlich wäre es möglich mit geringen Anpassungen auch Tonuino darauf laufen zu lassen, sobald man einen RFID-Reader integriert.

z.B. der Output von SPDIF-Signalen über die I2S-Schnittstelle und das Abspielen von Streams im Netzwerk vom Logitech Media Server sind interessant.

https://github.com/sle118/squeezelite-esp32
https://github.com/rochuck/squeeze-amp-too


---------------------------------------
I2S mit MCLK:

hierzu weiß ich nicht wozu man das braucht, aber hier ist beschrieben, wie man mit dem ESP32 ein I2S-Signal mit MCLK machen kann.
https://github.com/YetAnotherElectronicsChannel/ESP32_DSP_I2S_SETUP/blob/master/Documentation.pdf



Viele Grüße,
tyllmoritz

**Reply by tyllmoritz** [reply]
Cool ist auch, dass der Squeeze-Amp-Too direkt einen ESP32-WROVER-B verwendet und alles auf einer Platine ist.

Gab es hier nicht schon so ähnliche Ideen?

**Reply by Soehm** [reply]
Ich bin derzeit dran, ein AiO Board auf Basis ESP32 WROOM bei easyeda zu zeichnen. Allerdings stellt mich die Kopfhörerplatine noch vor eine gewisse Herausforderung - ich würde gerne auf den MAX98357 (und PCM5102+TDA1308) verzichten und alles über einen Chip machen. In meiner Vorstellung sollte der TLV320DAC3100 das bewerkstelligen, aber zumindest bei jlcpcb ist der nicht in Stock.

Punkt 2 ist sicherlich der Preis. Stand jetzt benötige ich viele extended Parts, was den Preis deutlich in die Höhe treibt und es unwirtschaftlich gegenüber einer "erweiterten Trägerplatine" mit Anschluss für PCM5102 + TDA1308 macht:
![Traegerplatine](https://user-images.githubusercontent.com/77161403/104467160-47168880-55b6-11eb-8130-2235563093da.jpg)

Bild der "unfertigen AiO" folgt, ich route noch...

Stefan

**Reply by biologist79** [reply]
Habe mal nach dem TLV320DAC3100 geschaut; der hat einen i2c-Anschluss. Für den muss man vermutlich einen Treiber schreiben, oder? Ansonsten zu deiner Platine zwei Fragen
a) Warum haben die Tasteranschlüsse drei Pins statt zwei?
b) Stellst du die SD-Karte senkrecht auf? Oder kommt der Kartenleser bei dir gar nicht aufs Board sondern woanders hin?

Habe meinen letzten PCB auch in blau bestellt. Ich mag ja blau :-)
Tja, man kriegt diese Einzelplatinen schon echt günstig. Muss man sich schon genau überlegen, ob sich das für einen lohnt...

Auf jeden Fall: Coole Idee, gefällt mir!

**Reply by Soehm** [reply]
Hey,

a) ich habe noch Schraubklemmen zu Hause, 3Pin. Der 3. Pin geht noch auf gpio's. Eventuell schreib ich den Code noch um, um die Led der Knöpfe anzusteuern. Bei IO00 weiß ich aber noch nicht, wie es sich verhält, daher auch die 3x 5V bzw GND
b) Hochkant bzw abgesetzt, um die SD von außen zu erreichen, mal schauen wie schnell der FTP ist...

Das Design ist noch nicht endgültig, aber die Flugverkabelung ist einfach zu fehleranfällig...

**Reply by Soehm** [reply]
TLV320DAC3100 - ja, Beschreibung inkl. Register liegt zum Studium auf dem Schreibtisch. Ist ja am Wochenende nix los, da hab ich Abends Zeit...

**Reply by biologist79** [reply]
FTP liegt über SPI bei 185 kB/s und via SD_MMC bei etwa 310 kB/s. SD_MMC spart auch nen GPIO, aber dafür ist man hinsichtlich der zu verwendenden GPIOs limitiert.

**Reply by compactflash** [reply]
Hallo , ich habe die Kopfhörerplatine entwickelt . Im Moment plane ich auch eine AiO , allerdings ohne fertige Module . Getestet ist SD_MMC (ok) und Audioteil (noch nicht final).
Ursprünglich wollte ich für den Tonuino32 als DAC den PT8211 verwenden , habe es aber verworfen weil dieser mit dem  MAX98357 nicht zusammen läuft . Jetzt einige Monate später bin ich schlauer und jetzt würde ich es so machen :
Getestet habe ich PT8211 ( Mini-Änderung in der Audio.cpp erforderlich ) mit TDA1308 und PAM8403 . Die Beschaltung ist einfach und alle Bauteile sind "große" SMD . Wie das mal aussehen wird und der TDA1308 evtl. durch einen anderen Baustein ersetzt wird oder ganz wegfällt weiß ich noch nicht . PAM8403 wäre Stereo , Mono oder Stereo  evtl. mit NS4150 , YX8002 oder ???? , alles möglich . 
Ich bleibe dran , habe aber im Moment aber wenig Zeit

**Comment by biologist79** [comment]
Die Grenzen sind hier halt fließend zwischen den Projekten. Ich habe bei mir im Gartenhaus beispielsweise einen Raspi, auf dem Volumio läuft. Da steckt eine ClassD-Amp von Hifiberry drauf mit 30W, die den Raspi gleich mitversorgt. Da gibt's auch zahlreiche Plugins. Da musste ich halt überhaupt keine Arbeit reinstecken; da flasht man Volumio auf eine SD, konfiguriert ein bisschen und fertig. Onlinedienste kannst damit auch nutzen, wobei ich für sowas allerdings keine Verwendung habe. 
In mein Tonuino-Projekt habe ich BT ja bereits "so ein bisschen" integriert. Also mal grundsätzlich läuft das, das Problem ist nur, dass es im Speicher langsam echt eng wird. Also mir schwebt vor, dass ich eine Modifikationskarte einführe, mit der man die Signalquelle auf BT umschalten kann (und wieder zurück). Um noch etwas Platz im Flashspeicher zu kriegen, würde ich die Webseiten, die aktuell ja im Flash sind, auf die SD auslagern. Aber ist alles noch Zukunftsmusik.

Das mit der eigenen Platine ist grundsätzlich ne gute Sache, nur ist das halt auch eine hohe Einstiegshürde. Dafür musst halt erstmal SMD-Löten können (was ich mir zutraue, aber selbst noch nie gemacht habe). Ich weiß noch nicht, ob ich das Ganze wirklich so weit treiben will, dass ich fertige Platinen hier anbiete. Weil grundsätzlich ist es ja auch so, dass die Develboards, wenn man sie nicht gerade in Deutschland kauft, echt ziemlich günstig sind. Mit Lolin32 bzw. Lolin D32 bin ich total zufrieden muss ich sagen, ich vermisse da nix. Der Lolin D32 pro hat halt einen WROVER an Bord, was den Vorteil bringt, dass man auch PSRAM hat. Mindestens mal fürs Webstreaming ist das gut, weil man das als Cache nutzen kann. Aber ob mir das weitergehend noch hilft in meinem Projekt, das weiß ich noch nicht. Möglich, dass es für Bluetooth auch gut wäre, aber das habe ich noch nicht ausgetestet.
Also kurzum: Ich möchte das nicht ausschließen, aber aktuell ist von mir keine eigene Platine geplant, die alle Bestandteile vereint. Aber der User, der die Kopfhörerplatine umgesetzt hat, zieht so etwas betracht. Was jetzt der aktuelle Stand ist weiß ich aber nicht.
